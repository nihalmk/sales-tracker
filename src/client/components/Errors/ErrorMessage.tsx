import React from 'react';
import { Alert } from 'tabler-react';

interface Props {
  error: string;
}
const ErrorMessage: React.FC<Props> = ({ error }) => (
  <React.Fragment>
    {error && (
      <Alert className="alert-align" type="danger">
        {error.replace('GraphQL error: ', '')}
      </Alert>
    )}
    <style jsx global>{``}</style>
  </React.Fragment>
);

export default ErrorMessage;
