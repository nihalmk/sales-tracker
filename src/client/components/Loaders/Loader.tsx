import React from 'react';
import { Logo } from '../Header/Logo';

const Loader = () => (
  <React.Fragment>
    <div className={'dimmer active p-5'}>
      <div className="loader">
        <Logo setColor />
      </div>
      <div className="dimmer-content">Loading...</div>
    </div>
    <style jsx global>{``}</style>
  </React.Fragment>
);

export default Loader;
