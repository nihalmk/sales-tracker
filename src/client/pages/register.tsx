import React from 'react';
import { NextPage } from 'next';
import { Layout } from '../components/Layout/Layout';
import Register from '../components/User/Register';

interface Props {}

const RegisterPage: NextPage<Props> = () => {

  return (
    <React.Fragment>
      <Layout hideHeader={false}><Register /></Layout>
    </React.Fragment>
  );
};

export default RegisterPage;
