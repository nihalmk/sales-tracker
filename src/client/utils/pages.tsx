export const Pages = {
  INDEX: '/',
  DASHBOARD: '/dashboard',
  LOGIN: '/login',
};

export const AuthenticatedPages: { [key: string]: string } = {
  DASHBOARD: '/dashboard',
  INDEX: '/',
};

export const Titles: { [key: string]: string } = {
  HOME: 'Sales Tracker',
  '/': 'Order Dashboard',
  '/dashboard': 'Order Dashboard',
  '/login': 'Login',
};
