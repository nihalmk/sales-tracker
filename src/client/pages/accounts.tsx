import React from 'react';
import { NextPage } from 'next';
import { Layout } from '../components/Layout/Layout';
import Link from 'next/link';
import { Pages } from '../utils/pages';

interface Props {}

const Accounts: NextPage<Props> = () => {
  return (
    <React.Fragment>
      <Layout hideHeader={false}>
        <div className="alert alert-primary" role="alert">
          Please send a mail to <strong>shoptrackerforsales@gmail.com</strong>{' '}
          to receive details regarding purchasing a paid account!
        </div>
        <Link href={Pages.DASHBOARD}>
          <button className="btn btn-primary mt-3">Go to home page</button>
        </Link>
      </Layout>
    </React.Fragment>
  );
};

export default Accounts;
