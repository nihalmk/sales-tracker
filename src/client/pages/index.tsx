import React, { useContext } from 'react';
import { useEffect } from 'react';
import { NextPage } from 'next';
import { clientLogger } from '../utils/logger';
import { useRouter } from 'next/router';
import { Pages } from '../utils/pages';
import Loader from '../components/Loaders/Loader';
import UserContext from '../components/UserWrapper/UserContext';
import AddShop from '../components/Shop/AddShop';
import { Layout } from '../components/Layout/Layout';

interface Props {}

const Home: NextPage<Props> = () => {
  const { user } = useContext(UserContext);

  const shop = user?.shop;

  const router = useRouter();
  useEffect(() => {
    if (shop) {
      clientLogger.log('redirecting to dashboard..');
      router.push(Pages.DASHBOARD);
    }
  }, [shop]);

  return (
    <React.Fragment>
      <Layout hideHeader={false}>{shop ? <Loader /> : <AddShop />}</Layout>
    </React.Fragment>
  );
};

export default Home;
