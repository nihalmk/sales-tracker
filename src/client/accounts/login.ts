import { accountsPassword, accountsClient } from '../../accounts/client';
import { LoginResult } from '@accounts/graphql-api';
import cookie from 'js-cookie';
import { clientLogger } from '../utils/logger';

export const login = async (
  email: string,
  password: string,
): Promise<LoginResult> => {
  const loginResult = await accountsPassword.login({
    password: password,
    user: { email: email },
  });
  cookie.set('token', loginResult.tokens.accessToken, { expires: 1 });
  return loginResult;
};

export const logout = async (): Promise<void> => {
  cookie.remove('token');
  await accountsClient.logout();
};

export const validateToken = async (): Promise<void> => {
  try {
    // Using refresh session to get token. Access token expiry is 90 minutes and refresh token expiry is 7days.
    // In case, access token is expired due to inactivity, refresh token will be used to generate new access token.
    const tokens = await accountsClient.refreshSession();

    if (!tokens || !tokens.accessToken) {
      logout();
      return;
    }
    const tokenFromCookie = cookie.get('token');

    // If token is not there or token is refreshed, add new token
    if (!tokenFromCookie || tokenFromCookie !== tokens.accessToken) {
      cookie.set('token', tokens.accessToken, { expires: 1 });
    }
  } catch (e) {
    // log the error.
    clientLogger.error(
      'Error while refreshing session. Error - ',
      JSON.stringify(e),
    );
  }
};
