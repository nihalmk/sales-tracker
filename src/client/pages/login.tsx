import Head from 'next/head';
import { useState, useEffect } from 'react';
import { login } from '../accounts/login';
import { useRouter } from 'next/router';
import { Pages } from '../utils/pages';
import { NextPage } from 'next';
import Loader from '../components/Loaders/Loader';
import ErrorMessage from '../components/Errors/ErrorMessage';
import { Logo } from '../components/Header/Logo';
import { GET_USER } from '../graphql/query/user';
import { useQuery } from '@apollo/react-hooks';
import Link from 'next/link';

interface Props {}

export const Login: NextPage<Props> = function () {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isLoading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const router = useRouter();

  const currentUser = useQuery(GET_USER, {
    fetchPolicy: 'no-cache',
  });

  const onEmailChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setEmail(event.target.value);
  };

  const onPasswordChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setPassword(event.target.value);
  };

  const onSubmit = async (e: React.SyntheticEvent) => {
    e && e.preventDefault();
    setLoading(true);
    try {
      if (!email || !password) {
        setError('Email/Password required');
        setLoading(false);
        return;
      }
      await login(email, password);
      router.push(Pages.INDEX);
    } catch (e) {
      setError(e.message);
      setLoading(false);
    }
  };

  useEffect(() => {
    if (currentUser?.data?.me) {
      router.push(Pages.DASHBOARD);
    }
  }, [currentUser]);

  if (currentUser.loading || currentUser?.data?.me) {
    return <Loader />;
  }
  return (
    <div className="container">
      <Head>
        <title>Login</title>
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <div className="row">
        <div className="col col-login mx-auto">
          <div className="text-center mb-6">
            <Logo setColor />
          </div>
          <form className="card" action="" method="post" onSubmit={onSubmit}>
            <div className="card-body p-6">
              <div className="card-title text-center">Sign in to your Shop</div>
            </div>
            <div className="card-body p-6">
              <ErrorMessage error={error}></ErrorMessage>
              <div className="form-group">
                <label className="form-label">Email address</label>
                <input
                  name="email"
                  type="email"
                  className="form-control"
                  id="email"
                  placeholder="Enter email"
                  onChange={onEmailChange}
                />
              </div>
              <div className="form-group">
                <label className="form-label">
                  Password
                  {/* <a className="float-right small" href="/forgot">
                    I forgot my password
                  </a> */}
                </label>
                <input
                  name="password"
                  type="password"
                  className="form-control"
                  id="userPassword"
                  placeholder="Password"
                  onChange={onPasswordChange}
                />
              </div>
              <Link href="/forgot">Forgot Password?</Link>
              <div className="form-footer">
                <button
                  type="button"
                  onClick={onSubmit}
                  className={
                    'btn btn-primary btn-block ' + (isLoading && 'btn-loading')
                  }
                >
                  Sign in
                </button>
                <button type="submit" hidden></button>
              </div>
            </div>
          </form>
        </div>
      </div>
      <style jsx>{``}</style>
    </div>
  );
};

export default Login;
