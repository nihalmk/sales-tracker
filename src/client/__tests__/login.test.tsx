import 'jsdom-global/register';
import React from 'react';
import { mount, shallow, ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Login } from '../pages/login';
import { login } from '../accounts/login';
import { user } from '../__stubs__/login.data';
import { createSerializer } from 'enzyme-to-json';
import Loader from '../components/Loaders/Loader';
import ErrorMessage from '../components/Errors/ErrorMessage';

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }) as any);

let mockUser: any;

let mockUserData: any = {
  data: {
    me: mockUser,
  },
};
jest.mock('../accounts/login');
jest.mock('next/router', () => {
  return {
    useRouter: jest.fn(() => ({
      push: jest.fn(),
    })),
  };
});

jest.mock('@apollo/react-hooks', () => ({
  __esModule: true,
  useQuery: (query: any): any => {
    if (!query) {
      return { data: {}, loading: false, error: true };
    }
    const mockQueryName = query.definitions[0].name.value;
    switch (mockQueryName) {
      case 'Me':
        return mockUserData;
      default: {
        return { data: null, error: true };
      }
    }
  },
}));

let wrapper: ReactWrapper;
const mockLogin = (login as unknown) as jest.Mock<typeof login> & typeof login;

describe('Login Page', () => {
  beforeAll(() => {
    mockUserData = {
      data: {
        me: null,
      },
    };
  });

  beforeEach(() => {
    mockLogin.mockClear();
  });
  it('Login page snapshot', () => {
    const snap = shallow(<Login />);
    expect(snap).toMatchSnapshot();
  });

  it('Login Page should be loaded and submitted for login', () => {
    wrapper = mount(<Login />);
    const email = wrapper.find('#email').first().props();
    const password = wrapper.find('#userPassword').first().props();
    const loginButton = wrapper.find('button').first();
    act(() => {
      email.onChange({
        target: {
          value: 'email@email.com',
        },
      } as React.ChangeEvent<HTMLInputElement>);
      password.onChange({
        target: {
          value: 'password',
        },
      } as React.ChangeEvent<HTMLInputElement>);
    });
    wrapper.update();
    act(() => {
      loginButton.simulate('click');
    });
    expect(mockLogin).toBeCalledWith('email@email.com', 'password');
  });

  it('Login Page should be loaded and should throw error for validation', () => {
    wrapper = mount(<Login />);
    const email = wrapper.find('#email').first().props();
    const loginButton = wrapper.find('button').first();
    act(() => {
      email.onChange({
        target: {
          value: 'email@email.com',
        },
      } as React.ChangeEvent<HTMLInputElement>);
    });
    wrapper.update();
    act(() => {
      loginButton.simulate('click');
    });
    expect(mockLogin).not.toBeCalledWith('email@email.com', 'password');
  });

  it('Login Page should be loaded and should throw error on login', async () => {
    wrapper = mount(<Login />);
    const email = wrapper.find('#email').first().props();
    const password = wrapper.find('#userPassword').first().props();
    const loginButton = wrapper.find('button').first();
    mockLogin.mockRejectedValueOnce(new Error('Invalid Credentials') as never);
    act(() => {
      email.onChange({
        target: {
          value: 'email@email.com',
        },
      } as React.ChangeEvent<HTMLInputElement>);
      password.onChange({
        target: {
          value: 'password',
        },
      } as React.ChangeEvent<HTMLInputElement>);
    });
    wrapper.update();
    await act(async () => {
      loginButton.simulate('click');
      await Promise.resolve();
    });
    wrapper.update();
    const error = wrapper.find(ErrorMessage).first();
    expect(mockLogin).toBeCalledWith('email@email.com', 'password');
    expect(error.html()).toContain('Invalid Credentials');
  });
});

describe('Login Page with user already logged in', () => {
  beforeAll(() => {
    mockUserData = {
      data: {
        me: user,
      },
    };
  });

  it('Login page snapshot', () => {
    const snap = shallow(<Login />);
    expect(snap).toMatchSnapshot();
  });

  it('Login Page should not be shown', () => {
    wrapper = mount(<Login />);
    const email = wrapper.find('#email').first();
    const password = wrapper.find('#userPassword').first();
    const loginButton = wrapper.find('button');
    const loader = wrapper.find(Loader);
    expect(email).toEqual({});
    expect(password).toEqual({});
    expect(loginButton).toEqual({});
    expect(loader).toBeTruthy();
  });
});
