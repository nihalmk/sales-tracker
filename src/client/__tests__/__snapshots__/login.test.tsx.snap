// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Login page snapshot 1`] = `
<div
  className="jsx-2085888330 container"
>
  <Head>
    <title
      className="jsx-2085888330"
    >
      Login
    </title>
    <link
      className="jsx-2085888330"
      href="/favicon.ico"
      rel="icon"
    />
  </Head>
  <div
    className="jsx-2085888330 row"
  >
    <div
      className="jsx-2085888330 col col-login mx-auto"
    >
      <div
        className="jsx-2085888330 text-center mb-6"
      >
        <Logo
          setColor={true}
        />
      </div>
      <form
        action=""
        className="jsx-2085888330 card"
        method="post"
      >
        <div
          className="jsx-2085888330 card-body p-6"
        >
          <div
            className="jsx-2085888330 card-title text-center"
          >
            Sign in to your Kitchen
          </div>
        </div>
        <div
          className="jsx-2085888330 card-body p-6"
        >
          <ErrorMessage
            error=""
          />
          <div
            className="jsx-2085888330 form-group"
          >
            <label
              className="jsx-2085888330 form-label"
            >
              Email address
            </label>
            <input
              className="jsx-2085888330 form-control"
              id="email"
              name="email"
              onChange={[Function]}
              placeholder="Enter email"
              type="email"
            />
          </div>
          <div
            className="jsx-2085888330 form-group"
          >
            <label
              className="jsx-2085888330 form-label"
            >
              Password
            </label>
            <input
              className="jsx-2085888330 form-control"
              id="userPassword"
              name="password"
              onChange={[Function]}
              placeholder="Password"
              type="password"
            />
          </div>
          <Link
            href="/forgot"
          >
            Forgot Password?
          </Link>
          <div
            className="jsx-2085888330 form-footer"
          >
            <button
              className="jsx-2085888330 btn btn-primary btn-block false"
              onClick={[Function]}
              type="button"
            >
              Sign in
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
  <JSXStyle
    id="2085888330"
  />
</div>
`;

exports[`Login Page with user already logged in Login page snapshot 1`] = `<Loader />`;
